Hello world Kotlin Jetpack Compose Example; Fixes:#3550

himanshumahajan138 opened the pull request:
# Pull Request 
Fixes: #3550 

## Description
Added `Kotlin Android Jetpack Compose Example` insuring proper integration of `AndroidAppKotlinModule` with `Jetpack Compose`

## Related Issues
- Link to related issue #3550.

## Checklist
- [x] Android hello world application with Jetpack Compose integration.
- [x] Code is clean and follows project conventions.
- [x] Documentation has been updated (to be done after review)

## Additional Notes
Not Fully Complete and Correct

lihaoyi commented on kotlinlib/src/mill/kotlinlib/android/AndroidAppKotlinModule.scala:
```diff
@@ -18,4 +21,66 @@ import mill.javalib.android.AndroidAppModule
+
+    aarFiles.map { aarFile =>
+      val extractDir = T.dest / aarFile.baseName
+      os.call(Seq("unzip", aarFile.toString, "-d", extractDir.toString))
+      PathRef(extractDir / "classes.jar")
```
This should use the `os.unzip` API

himanshumahajan138 replied:
done

lihaoyi commented on kotlinlib/src/mill/kotlinlib/android/AndroidAppKotlinModule.scala:
```diff
@@ -18,4 +21,66 @@ import mill.javalib.android.AndroidAppModule
+   *
+   * @return A task yielding the updated classpath with `.jar` files.
+   */
+  def updatedCompileClasspath: T[Agg[PathRef]] = Task {
+    super.compileClasspath().filter(_.path.ext == "jar") ++ Agg.from(recompileAARs())
+  }
```
This can be done directly in `compileClasspath` and doesnt need a separate task

himanshumahajan138 replied:
actually i tried but it was getting stucked in loop and give stackoverflow error

lihaoyi replied:
It shouldn't, you'll need to figure out why it's misbehaving

himanshumahajan138 replied:
done

lihaoyi commented on kotlinlib/src/mill/kotlinlib/android/AndroidAppKotlinModule.scala:
```diff
@@ -18,4 +21,66 @@ import mill.javalib.android.AndroidAppModule
+   *
+   * @return A sequence of `PathRef` pointing to the extracted JAR files.
+   */
+  def recompileAARs: T[Seq[PathRef]] = Task {
```
This needs a better name. Maybe `androidUnpackArchives`?

himanshumahajan138 replied:
ya it is i'll update

lihaoyi commented on kotlinlib/src/mill/kotlinlib/android/AndroidAppKotlinModule.scala:
```diff
@@ -18,4 +21,66 @@ import mill.javalib.android.AndroidAppModule
+   * Implicit `ReadWriter` for serializing and deserializing `coursier.Type` values.
+   * Converts a `coursier.Type` to a `String` and vice versa.
+   */
+  implicit val coursierTypeRW: ReadWriter[Type] = upickle.default.readwriter[String].bimap(
+    _.value, // Serialize coursier.Type to String
+    Type(_) // Deserialize String to coursier.Type
+  )
+
+  /**
+   * Implicit `ReadWriter` for handling `Set[coursier.Type]`, allowing conversion
+   * between a set of `coursier.Type` and a set of `String`.
+   */
+  implicit val coursierTypeSetRW: ReadWriter[Set[Type]] =
+    upickle.default.readwriter[Set[String]].bimap(
+      _.map(_.value), // Serialize Set[coursier.Type] to Set[String]
+      _.map(Type(_)) // Deserialize Set[String] to Set[coursier.Type]
+    )
```
These should go into `JsonFormatters.scala`

himanshumahajan138 replied:
ok and then do we need to import it separately or it will get itself

lihaoyi replied:
Everything in `JsonFormatters.scala` gets pulled in by `import mill._`

himanshumahajan138 replied:
done

himanshumahajan138 replied:
done

lihaoyi commented on kotlinlib/src/mill/kotlinlib/android/AndroidAppKotlinModule.scala:
```diff
@@ -18,4 +21,66 @@ import mill.javalib.android.AndroidAppModule
  */
 @mill.api.experimental
-trait AndroidAppKotlinModule extends AndroidAppModule with KotlinModule {}
+trait AndroidAppKotlinModule extends AndroidAppModule with KotlinModule {
```
Most of this stuff should probably go into `AndroidAppModule` since non-kotlin android apps will need it too

himanshumahajan138 replied:
actually not because we have to implement this for jetpack compose 

my thoughts
if we do this in same AndroidAppModule may be it will get some conflict with java and kt both...

lihaoyi replied:
The point of this task is not to make it support jetpack compose specifically, but to use jetpack compose as an example with requirements that all android apps will need. Nothing here seems specific to Kotlin apps so we should not limit it

himanshumahajan138 replied:
done

lihaoyi commented:
Left some comments. Looks reasonable overall. Haven't tested it manually, but will do before merging.

himanshumahajan138 commented:
All Reviews Has been Resolved if any further changes are needed then please review this again 

If Code is done, then I will move to the documentation part (please confirm once)

lihaoyi commented:
Please proceed with documentation, you do not meed to wait until code is done

0xnm commented:
I think what you are missing here (to be precise - in your previous PR) is support of the resources: drawables (at least for the icon), strings, etc. Because no real-world app won't be released without those. There is also a question of different qualifiers support, but this is another story.

himanshumahajan138 commented:
> I think what you are missing here (to be precise - in your previous PR) is support of the resources: drawables (at least for the icon), strings, etc. Because no real-world app won't be released without those. There is also a question of different qualifiers support, but this is another story.

yes Sir, i know actually when i raised this pr i just solved the unresolved reference error and i think that its working but today when i tested my android application then i came to know that its crashing(and discussed same in discord about aar full support today)

Now i am working in the rest of the implementation and if you know something about how the android aar files are handled in the backend(like how their resources are merged, manifest.xml merge, R.txt merge and many other things) i my self have reached out a lot of android developers but still didn't got the right answer 

no worries i will try my best to solve this issue 

any help appreciated...

himanshumahajan138 commented:
Finally, its Done App is running and Working 

Documentation also Updated; All Tests Passed 

@lihaoyi Please Provide Final reviews if Required....

Please Pardon me if i made any mistake...

lihaoyi commented:
@himanshumahajan138 I tried this out locally, it appears that we do not make any use of Jetpack Compose in the actual application example, and everything still works even when the Compose ivydeps and imports have been commented out. This needs to use Jetpack Compose to in the application to make sure it actually works

himanshumahajan138 commented:
@lihaoyi I Will make it work also but first i think we should do something for SDK coz downloading it again and again sense out nothing

himanshumahajan138 commented:
@lihaoyi Sir I have Updated whole android Setup, Examples and fixed that error (if ivydeps are commented)

Please Review this....

Pardon me if any mistake done by me...

lihaoyi commented:
@himanshumahajan138 example/kotlinlib/android/2-jetpack-compose-hello-world/app/src/java/com/helloworld/app/MainActivity.kt still does not make use of Jetpack compose in its implementation.

Please take the source code from https://github.com/android/codelab-android-compose/blob/main/BasicsCodelab/app/src/main/java/com/codelab/basics/MainActivity.kt and ensure it can build and run correctly

himanshumahajan138 commented:
@lihaoyi sorry sir its my bad, I will update the code...
Thanks for being cooperative...
